File: apps/windows/src/services/pairingClient.ts
import type {
  DeviceIdentity,
  PairingCompleteMessage,
  PairingExpiredMessage,
  PairingJoinedMessage,
  PairingQrPayload,
  PairingSessionCreatedMessage,
  PairingViewState,
  TrustedDevice
} from "../types/pairing.js";
import { loadOrCreateWindowsIdentity } from "./identityStore.js";
import { RelayClient } from "./relayClient.js";
import { saveTrustedDevice } from "./trustedDeviceStore.js";

type PairingStateHandler = (state: PairingViewState) => void;

const RELAY_HELLO_TIMEOUT_MS = 8_000;

function randomBytes(length: number): Uint8Array {
  const bytes = new Uint8Array(length);
  if (globalThis.crypto?.getRandomValues) {
    globalThis.crypto.getRandomValues(bytes);
    return bytes;
  }

  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Math.floor(Math.random() * 256);
  }
  return bytes;
}

function bytesToBase64(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) {
    binary += String.fromCharCode(byte);
  }
  return btoa(binary);
}

function randomToken(prefix: string, byteLength = 18): string {
  return `${prefix}_${bytesToBase64(randomBytes(byteLength))
    .replaceAll("+", "-")
    .replaceAll("/", "_")
    .replaceAll("=", "")}`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isDeviceIdentity(value: unknown): value is DeviceIdentity {
  return isRecord(value) &&
    typeof value.deviceId === "string" &&
    typeof value.deviceName === "string" &&
    (value.platform === "windows" || value.platform === "android") &&
    typeof value.publicKey === "string";
}

function isQrPayload(value: unknown): value is PairingQrPayload {
  return isRecord(value) &&
    value.protocol === "crossbridge-v1" &&
    typeof value.pairingSessionId === "string" &&
    typeof value.relayUrl === "string" &&
    typeof value.pcDeviceId === "string" &&
    typeof value.pcDeviceName === "string" &&
    typeof value.pcPublicKey === "string" &&
    typeof value.pairingToken === "string" &&
    typeof value.expiresAt === "number";
}

function isTrustedDevice(value: unknown): value is TrustedDevice {
  if (!isRecord(value) || !isDeviceIdentity(value)) return false;
  const record = value as Record<string, unknown>;
  return typeof record.pairedAt === "number" &&
    (record.lastSeenAt === undefined || typeof record.lastSeenAt === "number");
}

function isMessageOfType<TType extends string>(
  message: unknown,
  type: TType
): message is { type: TType; payload?: unknown } {
  return isRecord(message) && message.type === type;
}

function isSessionCreatedMessage(message: unknown): message is PairingSessionCreatedMessage {
  return isMessageOfType(message, "PAIRING_SESSION_CREATED") &&
    isRecord(message.payload) &&
    isQrPayload(message.payload.qrPayload);
}

function isJoinedMessage(message: unknown): message is PairingJoinedMessage {
  return isMessageOfType(message, "PAIRING_JOINED") &&
    isRecord(message.payload) &&
    typeof message.payload.pairingSessionId === "string" &&
    isDeviceIdentity(message.payload.pcIdentity) &&
    isDeviceIdentity(message.payload.androidIdentity) &&
    typeof message.payload.verificationCode === "string";
}

function isCompleteMessage(message: unknown): message is PairingCompleteMessage {
  return isMessageOfType(message, "PAIRING_COMPLETE") &&
    isRecord(message.payload) &&
    typeof message.payload.pairingSessionId === "string" &&
    Array.isArray(message.payload.trustedDevices) &&
    message.payload.trustedDevices.every(isTrustedDevice);
}

function isExpiredMessage(message: unknown): message is PairingExpiredMessage {
  return isMessageOfType(message, "PAIRING_EXPIRED") &&
    isRecord(message.payload) &&
    typeof message.payload.pairingSessionId === "string" &&
    typeof message.payload.expiresAt === "number";
}

function isRelayWelcome(message: unknown, deviceId: string): boolean {
  return isRecord(message) &&
    message.type === "RELAY_WELCOME" &&
    message.deviceId === deviceId &&
    message.protocolVersion === 1;
}

function isRelayError(message: unknown): message is { type: "RELAY_ERROR"; message: string } {
  return isRecord(message) &&
    message.type === "RELAY_ERROR" &&
    typeof message.message === "string";
}

export class PairingClient {
  private readonly relayClient: RelayClient;
  private readonly unsubscribeRelayMessages: () => void;
  private readonly unsubscribeRelayState: () => void;
  private readonly handlers = new Set<PairingStateHandler>();
  private pcIdentity?: DeviceIdentity;
  private state: PairingViewState;
  private currentPairingSessionId?: string;
  private pendingQrRelayUrl?: string;
  private relayWelcomeResolver?: () => void;
  private relayWelcomeRejecter?: (error: Error) => void;
  private expiryTimer?: ReturnType<typeof setTimeout>;

  constructor(relayClient = new RelayClient()) {
    this.relayClient = relayClient;
    this.state = {
      state: "idle",
      relayConnected: false,
      relayConnectionState: "disconnected"
    };
    this.unsubscribeRelayMessages = this.relayClient.onMessage((message) => {
      this.handleRelayMessage(message);
    });
    this.unsubscribeRelayState = this.relayClient.onStateChange((relayState) => {
      this.updateState({
        relayConnected: relayState === "connected",
        relayConnectionState: relayState
      });
    });
  }

  getState(): PairingViewState {
    return this.state;
  }

  onStateChange(handler: PairingStateHandler): () => void {
    this.handlers.add(handler);
    handler(this.state);
    return () => {
      this.handlers.delete(handler);
    };
  }

  async createPairingSession(relayUrl: string, qrRelayUrl?: string): Promise<void> {
    this.clearExpiryTimer();
    this.currentPairingSessionId = undefined;
    this.pendingQrRelayUrl = qrRelayUrl?.trim() || undefined;
    this.updateState({
      state: "connecting",
      relayConnected: false,
      qrPayload: undefined,
      verificationCode: undefined,
      androidIdentity: undefined,
      trustedAndroidDevice: undefined,
      error: undefined,
      expiresAt: undefined
    });

    try {
      const pcIdentity = await this.ensureWindowsIdentity();
      await this.relayClient.connect(relayUrl.trim());
      await this.sendRelayHello(pcIdentity);
      this.relayClient.send({
        type: "PAIRING_SESSION_CREATE",
        payload: {
          deviceIdentity: pcIdentity
        }
      });
    } catch (error) {
      this.updateState({
        state: "error",
        error: error instanceof Error ? error.message : "Could not create pairing session."
      });
    }
  }

  confirmPairing(): void {
    if (!this.currentPairingSessionId || !this.pcIdentity) return;
    try {
      this.relayClient.send({
        type: "PAIRING_CONFIRM",
        payload: {
          pairingSessionId: this.currentPairingSessionId,
          deviceId: this.pcIdentity.deviceId
        }
      });
      this.updateState({ state: "confirmed", error: undefined });
    } catch (error) {
      this.updateState({
        state: "error",
        error: error instanceof Error ? error.message : "Could not confirm pairing."
      });
    }
  }

  disconnect(): void {
    this.clearExpiryTimer();
    this.relayClient.disconnect();
  }

  dispose(): void {
    this.disconnect();
    this.unsubscribeRelayMessages();
    this.unsubscribeRelayState();
    this.handlers.clear();
  }

  private async ensureWindowsIdentity(): Promise<DeviceIdentity> {
    if (this.pcIdentity) return this.pcIdentity;
    const identity = await loadOrCreateWindowsIdentity();
    this.pcIdentity = identity;
    this.updateState({ pcIdentity: identity });
    return identity;
  }

  private sendRelayHello(pcIdentity: DeviceIdentity): Promise<void> {
    this.relayClient.send({
      type: "RELAY_HELLO",
      deviceId: pcIdentity.deviceId,
      sessionToken: randomToken("session", 18),
      protocolVersion: 1
    });

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.relayWelcomeResolver = undefined;
        this.relayWelcomeRejecter = undefined;
        reject(new Error("Relay did not acknowledge this Windows device."));
      }, RELAY_HELLO_TIMEOUT_MS);

      this.relayWelcomeResolver = () => {
        clearTimeout(timeout);
        this.relayWelcomeResolver = undefined;
        this.relayWelcomeRejecter = undefined;
        resolve();
      };
      this.relayWelcomeRejecter = (error) => {
        clearTimeout(timeout);
        this.relayWelcomeResolver = undefined;
        this.relayWelcomeRejecter = undefined;
        reject(error);
      };
    });
  }

  private handleRelayMessage(message: unknown): void {
    if (this.pcIdentity && isRelayWelcome(message, this.pcIdentity.deviceId)) {
      this.relayWelcomeResolver?.();
      return;
    }

    if (isRelayError(message)) {
      this.relayWelcomeRejecter?.(new Error(message.message));
      this.updateState({ state: "error", error: message.message });
      return;
    }

    if (isSessionCreatedMessage(message)) {
      const qrPayload = this.pendingQrRelayUrl
        ? { ...message.payload.qrPayload, relayUrl: this.pendingQrRelayUrl }
        : message.payload.qrPayload;
      this.currentPairingSessionId = qrPayload.pairingSessionId;
      this.scheduleExpiry(qrPayload);
      this.updateState({
        state: "waiting_for_android",
        qrPayload,
        expiresAt: qrPayload.expiresAt,
        error: undefined
      });
      return;
    }

    if (isJoinedMessage(message) && message.payload.pairingSessionId === this.currentPairingSessionId) {
      this.updateState({
        state: "waiting_for_confirmation",
        androidIdentity: message.payload.androidIdentity,
        verificationCode: message.payload.verificationCode,
        error: undefined
      });
      return;
    }

    if (isCompleteMessage(message) && message.payload.pairingSessionId === this.currentPairingSessionId) {
      this.clearExpiryTimer();
      void this.saveCompletedTrustedAndroidDevice(message.payload.trustedDevices);
      return;
    }

    if (isExpiredMessage(message) && message.payload.pairingSessionId === this.currentPairingSessionId) {
      this.clearExpiryTimer();
      this.updateState({
        state: "expired",
        expiresAt: message.payload.expiresAt
      });
    }
  }

  private async saveCompletedTrustedAndroidDevice(trustedDevices: TrustedDevice[]): Promise<void> {
    if (!this.pcIdentity) {
      this.updateState({
        state: "error",
        error: "Windows identity was not available when pairing completed."
      });
      return;
    }

    const trustedAndroidDevice = trustedDevices.find((device) => {
      return device.platform === "android" &&
        device.deviceId !== this.pcIdentity?.deviceId;
    });

    if (trustedAndroidDevice) {
      const deviceWithLastSeen = {
        ...trustedAndroidDevice,
        lastSeenAt: trustedAndroidDevice.lastSeenAt ?? Date.now()
      };
      try {
        await saveTrustedDevice(deviceWithLastSeen);
        this.updateState({
          state: "complete",
          trustedAndroidDevice: deviceWithLastSeen,
          error: undefined
        });
      } catch (error) {
        this.updateState({
          state: "error",
          error: error instanceof Error
            ? error.message
            : "Pairing completed, but the trusted device could not be saved."
        });
      }
    } else {
      this.updateState({
        state: "complete",
        error: undefined
      });
    }
  }

  private scheduleExpiry(qrPayload: PairingQrPayload): void {
    this.clearExpiryTimer();
    const delayMs = Math.max(0, qrPayload.expiresAt - Date.now());
    this.expiryTimer = setTimeout(() => {
      if (this.currentPairingSessionId !== qrPayload.pairingSessionId) return;
      if (this.state.state === "complete" || this.state.state === "confirmed") return;
      this.updateState({ state: "expired", expiresAt: qrPayload.expiresAt });
    }, delayMs);
  }

  private clearExpiryTimer(): void {
    if (this.expiryTimer) {
      clearTimeout(this.expiryTimer);
      this.expiryTimer = undefined;
    }
  }

  private updateState(nextState: Partial<PairingViewState>): void {
    this.state = {
      ...this.state,
      ...nextState
    };
    for (const handler of this.handlers) {
      handler(this.state);
    }
  }
}

File: apps/android/app/src/main/java/dev/crossbridge/android/network/PairingClient.kt
package dev.crossbridge.android.network

import android.content.Context
import dev.crossbridge.android.data.AndroidIdentityStore
import dev.crossbridge.android.data.RelayUrlStore
import dev.crossbridge.android.data.TrustedDeviceStore
import dev.crossbridge.android.protocol.DeviceIdentity
import dev.crossbridge.android.protocol.PairingCode
import dev.crossbridge.android.protocol.PairingMessageType
import dev.crossbridge.android.protocol.PairingQrPayload
import dev.crossbridge.android.protocol.TrustedDevice
import java.util.UUID
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

enum class PairingState {
    IDLE,
    QR_SCANNED,
    CONNECTING,
    JOIN_SENT,
    WAITING_FOR_WINDOWS,
    VERIFICATION_READY,
    CONFIRMED,
    COMPLETE,
    EXPIRED,
    ERROR
}

data class PairingViewState(
    val state: PairingState,
    val relayConnected: Boolean = false,
    val qrPayload: PairingQrPayload? = null,
    val verificationCode: String? = null,
    val androidIdentity: DeviceIdentity? = null,
    val pcIdentity: DeviceIdentity? = null,
    val error: String? = null,
    val expiresAt: Long? = null
)

class PairingClient(
    context: Context,
    private val relayClient: RelayClient = RelayClient()
) {
    private val identityStore = AndroidIdentityStore(context)
    private val relayUrlStore = RelayUrlStore(context)
    private val trustedDeviceStore = TrustedDeviceStore(context)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val _viewState = MutableStateFlow(PairingViewState(PairingState.IDLE))
    private val unsubscribeRelayMessages: () -> Unit
    private val unsubscribeRelayState: () -> Unit

    val viewState: StateFlow<PairingViewState> = _viewState.asStateFlow()

    private var currentPayload: PairingQrPayload? = null
    private var currentAndroidIdentity: DeviceIdentity? = null
    private var relayWelcome: CompletableDeferred<Unit>? = null
    private var expiryJob: Job? = null

    init {
        unsubscribeRelayMessages = relayClient.onMessage(::handleRelayMessage)
        unsubscribeRelayState = relayClient.onStateChange { relayState ->
            _viewState.update {
                it.copy(relayConnected = relayState == RelayConnectionState.CONNECTED)
            }
        }
    }

    fun startPairing(qrPayload: PairingQrPayload) {
        expiryJob?.cancel()
        currentPayload = qrPayload
        _viewState.value = PairingViewState(
            state = PairingState.QR_SCANNED,
            qrPayload = qrPayload,
            expiresAt = qrPayload.expiresAt
        )

        scope.launch {
            if (qrPayload.expiresAt <= System.currentTimeMillis()) {
                updateState(
                    state = PairingState.EXPIRED,
                    error = "Pairing code has expired. Create a new code on Windows."
                )
                return@launch
            }

            scheduleExpiry(qrPayload)
            val androidIdentity = identityStore.loadOrCreateIdentity()
            currentAndroidIdentity = androidIdentity
            updateState(
                state = PairingState.CONNECTING,
                androidIdentity = androidIdentity,
                error = null
            )

            try {
                relayClient.connect(qrPayload.relayUrl)
                sendRelayHello(androidIdentity)
                sendPairingJoin(qrPayload, androidIdentity)
                updateState(state = PairingState.WAITING_FOR_WINDOWS, error = null)
            } catch (error: Throwable) {
                updateState(
                    state = PairingState.ERROR,
                    error = error.message ?: "Could not join relay pairing."
                )
            }
        }
    }

    fun confirmPairing() {
        val payload = currentPayload ?: return
        val androidIdentity = currentAndroidIdentity ?: return
        val currentState = viewState.value
        if (currentState.state != PairingState.VERIFICATION_READY) return

        scope.launch {
            try {
                relayClient.send(
                    JSONObject()
                        .put("type", PairingMessageType.PAIRING_CONFIRM.wireValue)
                        .put(
                            "payload",
                            JSONObject()
                                .put("pairingSessionId", payload.pairingSessionId)
                                .put("deviceId", androidIdentity.deviceId)
                        )
                )
                updateState(state = PairingState.CONFIRMED, error = null)
            } catch (error: Throwable) {
                updateState(
                    state = PairingState.ERROR,
                    error = error.message ?: "Could not confirm pairing."
                )
            }
        }
    }

    fun reset() {
        expiryJob?.cancel()
        relayClient.disconnect()
        currentPayload = null
        currentAndroidIdentity = null
        relayWelcome = null
        _viewState.value = PairingViewState(PairingState.IDLE)
    }

    fun dispose() {
        expiryJob?.cancel()
        unsubscribeRelayMessages()
        unsubscribeRelayState()
        relayClient.close()
        scope.cancel()
    }

    private suspend fun sendRelayHello(androidIdentity: DeviceIdentity) {
        val welcome = CompletableDeferred<Unit>()
        relayWelcome = welcome
        relayClient.send(
            JSONObject()
                .put("type", "RELAY_HELLO")
                .put("deviceId", androidIdentity.deviceId)
                .put("sessionToken", "session_${UUID.randomUUID().toString().replace("-", "")}")
                .put("protocolVersion", 1)
        )
        try {
            withTimeout(RELAY_HELLO_TIMEOUT_MS) {
                welcome.await()
            }
        } finally {
            if (relayWelcome == welcome) {
                relayWelcome = null
            }
        }
    }

    private suspend fun sendPairingJoin(
        qrPayload: PairingQrPayload,
        androidIdentity: DeviceIdentity
    ) {
        updateState(state = PairingState.JOIN_SENT, error = null)
        relayClient.send(
            JSONObject()
                .put("type", PairingMessageType.PAIRING_JOIN.wireValue)
                .put(
                    "payload",
                    JSONObject()
                        .put("pairingSessionId", qrPayload.pairingSessionId)
                        .put("pairingToken", qrPayload.pairingToken)
                        .put("deviceIdentity", androidIdentity.toJson())
                )
        )
    }

    private fun handleRelayMessage(message: JSONObject) {
        when (message.optString("type")) {
            "RELAY_WELCOME" -> handleRelayWelcome(message)
            "RELAY_ERROR", PairingMessageType.ERROR.wireValue -> handleRelayError(message)
            PairingMessageType.PAIRING_JOINED.wireValue -> handlePairingJoined(message)
            PairingMessageType.PAIRING_COMPLETE.wireValue -> handlePairingComplete(message)
            PairingMessageType.PAIRING_EXPIRED.wireValue -> handlePairingExpired(message)
        }
    }

    private fun handleRelayWelcome(message: JSONObject) {
        val androidIdentity = currentAndroidIdentity ?: return
        if (message.optString("deviceId") == androidIdentity.deviceId) {
            relayWelcome?.complete(Unit)
        }
    }

    private fun handleRelayError(message: JSONObject) {
        val messageText = message.optString("message").ifBlank {
            message.optJSONObject("payload")?.optString("message").orEmpty()
        }.ifBlank {
            "Relay returned an error."
        }
        relayWelcome?.completeExceptionally(IllegalStateException(messageText))
        updateState(state = PairingState.ERROR, error = messageText)
    }

    private fun handlePairingJoined(message: JSONObject) {
        val payload = message.optJSONObject("payload") ?: return
        val qrPayload = currentPayload ?: return
        if (payload.optString("pairingSessionId") != qrPayload.pairingSessionId) return

        val pcIdentity = payload.optJSONObject("pcIdentity")?.toDeviceIdentity()
        val androidIdentity = payload.optJSONObject("androidIdentity")?.toDeviceIdentity()
        val verificationCode = payload.optString("verificationCode")
        if (pcIdentity == null || androidIdentity == null || !verificationCode.matches(CODE_REGEX)) {
            updateState(
                state = PairingState.ERROR,
                error = "Relay pairing response was invalid."
            )
            return
        }

        if (
            pcIdentity.deviceId != qrPayload.pcDeviceId ||
            pcIdentity.publicKey != qrPayload.pcPublicKey
        ) {
            updateState(
                state = PairingState.ERROR,
                error = "Windows identity from relay did not match the QR payload."
            )
            return
        }

        val expectedCode = PairingCode.derive(
            pcPublicKey = pcIdentity.publicKey,
            androidPublicKey = androidIdentity.publicKey,
            pairingSessionId = qrPayload.pairingSessionId
        )
        if (verificationCode != expectedCode) {
            updateState(
                state = PairingState.ERROR,
                error = "Verification code did not match the CrossBridge pairing calculation."
            )
            return
        }

        currentAndroidIdentity = androidIdentity
        updateState(
            state = PairingState.VERIFICATION_READY,
            verificationCode = verificationCode,
            androidIdentity = androidIdentity,
            pcIdentity = pcIdentity,
            error = null
        )
    }

    private fun handlePairingComplete(message: JSONObject) {
        val payload = message.optJSONObject("payload") ?: return
        val qrPayload = currentPayload ?: return
        if (payload.optString("pairingSessionId") != qrPayload.pairingSessionId) return

        val trustedDevices = payload.optJSONArray("trustedDevices").toTrustedDevices()
        val androidIdentity = currentAndroidIdentity
        val windowsDevice = trustedDevices.firstOrNull { trustedDevice ->
            trustedDevice.platform == "windows" &&
                trustedDevice.deviceId != androidIdentity?.deviceId
        }

        if (windowsDevice == null) {
            updateState(
                state = PairingState.ERROR,
                error = "Pairing completed, but Windows trusted-device metadata was missing."
            )
            return
        }

        trustedDeviceStore.saveTrustedDevice(windowsDevice)
        relayUrlStore.saveRelayUrl(qrPayload.relayUrl)
        expiryJob?.cancel()
        updateState(
            state = PairingState.COMPLETE,
            pcIdentity = DeviceIdentity(
                deviceId = windowsDevice.deviceId,
                deviceName = windowsDevice.deviceName,
                platform = windowsDevice.platform,
                publicKey = windowsDevice.publicKey
            ),
            error = null
        )
    }

    private fun handlePairingExpired(message: JSONObject) {
        val payload = message.optJSONObject("payload") ?: return
        val qrPayload = currentPayload ?: return
        if (payload.optString("pairingSessionId") != qrPayload.pairingSessionId) return
        expiryJob?.cancel()
        updateState(
            state = PairingState.EXPIRED,
            expiresAt = payload.optLong("expiresAt", qrPayload.expiresAt),
            error = "Pairing code has expired. Create a new code on Windows."
        )
    }

    private fun scheduleExpiry(qrPayload: PairingQrPayload) {
        expiryJob?.cancel()
        expiryJob = scope.launch {
            delay((qrPayload.expiresAt - System.currentTimeMillis()).coerceAtLeast(0))
            if (currentPayload?.pairingSessionId != qrPayload.pairingSessionId) return@launch
            if (viewState.value.state != PairingState.COMPLETE) {
                updateState(
                    state = PairingState.EXPIRED,
                    error = "Pairing code has expired. Create a new code on Windows."
                )
            }
        }
    }

    private fun updateState(
        state: PairingState? = null,
        relayConnected: Boolean? = null,
        qrPayload: PairingQrPayload? = null,
        verificationCode: String? = null,
        androidIdentity: DeviceIdentity? = null,
        pcIdentity: DeviceIdentity? = null,
        error: String? = null,
        expiresAt: Long? = null
    ) {
        _viewState.update { current ->
            current.copy(
                state = state ?: current.state,
                relayConnected = relayConnected ?: current.relayConnected,
                qrPayload = qrPayload ?: current.qrPayload,
                verificationCode = verificationCode ?: current.verificationCode,
                androidIdentity = androidIdentity ?: current.androidIdentity,
                pcIdentity = pcIdentity ?: current.pcIdentity,
                error = error,
                expiresAt = expiresAt ?: current.expiresAt
            )
        }
    }

    private fun DeviceIdentity.toJson(): JSONObject {
        return JSONObject()
            .put("deviceId", deviceId)
            .put("deviceName", deviceName)
            .put("platform", platform)
            .put("publicKey", publicKey)
    }

    private fun JSONObject.toDeviceIdentity(): DeviceIdentity? {
        val deviceId = optString("deviceId").trim().takeIf { it.isNotBlank() } ?: return null
        val deviceName = optString("deviceName").trim().takeIf { it.isNotBlank() } ?: return null
        val platform = optString("platform").trim().takeIf { it == "windows" || it == "android" }
            ?: return null
        val publicKey = optString("publicKey").trim().takeIf { it.isNotBlank() } ?: return null

        return DeviceIdentity(
            deviceId = deviceId,
            deviceName = deviceName,
            platform = platform,
            publicKey = publicKey
        )
    }

    private fun JSONArray?.toTrustedDevices(): List<TrustedDevice> {
        if (this == null) return emptyList()
        val devices = mutableListOf<TrustedDevice>()
        for (index in 0 until length()) {
            val device = optJSONObject(index)?.toTrustedDevice() ?: continue
            devices.add(device)
        }
        return devices
    }

    private fun JSONObject.toTrustedDevice(): TrustedDevice? {
        val deviceId = optString("deviceId").trim().takeIf { it.isNotBlank() } ?: return null
        val deviceName = optString("deviceName").trim().takeIf { it.isNotBlank() } ?: return null
        val platform = optString("platform").trim().takeIf { it == "windows" || it == "android" }
            ?: return null
        val publicKey = optString("publicKey").trim().takeIf { it.isNotBlank() } ?: return null
        val pairedAt = optLongOrNull("pairedAt") ?: return null
        val lastSeenAt = optLongOrNull("lastSeenAt")

        return TrustedDevice(
            deviceId = deviceId,
            deviceName = deviceName,
            platform = platform,
            publicKey = publicKey,
            pairedAt = pairedAt,
            lastSeenAt = lastSeenAt
        )
    }

    private fun JSONObject.optLongOrNull(field: String): Long? {
        if (!has(field) || isNull(field)) return null
        return try {
            getLong(field)
        } catch (_: JSONException) {
            null
        }
    }

    private companion object {
        const val RELAY_HELLO_TIMEOUT_MS = 8_000L
        val CODE_REGEX = Regex("^\\d{6}$")
    }
}
